### ISSUE TITLE
ctags doesn't find subsequent functions when a macro defines an else part for an if

### ISSUE THREAD

**Issue by Smjert** [universal-ctags/ctags#1188]
Hello, i have this short example (test.c):

```
int main()
{
        if (something) {
                printf("hello");
#ifdef world
        } else
                printf(" world\n");
#else
        }
#endif
}

int func()
{
}
```

Calling 'ctags test.c' only generate the tag for main() function but not for func().

Ctags Info:

> Universal Ctags 0.0.0(0b1b759), Copyright (C) 2015 Universal Ctags Team
> Universal Ctags is derived from Exuberant Ctags.
> Exuberant Ctags 5.8, Copyright (C) 1996-2009 Darren Hiebert
>   Compiled: Nov 14 2016, 17:05:27
>   URL: https://ctags.io/
>   Optional compiled features: +wildcards, +regex, +option-directory, +coproc

**Comment by b4n** [comment]
You example code is invalid if `world` is defined, as it results in unbalanced braces.  Use `} else {` for valid syntax in all branches.

**Comment by b4n** [comment]
…and actually your edited example is invalid in any case, still unbalanced braces.

**Comment by Smjert** [comment]
I'm sorry i did a mess when copying (there were spaces everywhere) and acidentally deleted the #else part.

**Comment by b4n** [comment]
Okay, now it's messy code, but it is valid alright.

@pragmaware @masatake OldC handles it OK (probably because it retries with another preprocessor config on unbalanced braces?)

**Comment by masatake** [comment]
This is the hardest issue in things ctags deals with.
I will never say this is a bug of parser. However, if ctags can handl these input well, people may be happy.

**Comment by b4n** [comment]
Agreed that I don't see it as a high priority thing.  It'd be nice to support, sure, but it's fairly tricky code for no reason.

I believe the OldC parser did a second pass with another option to `lcpp` when it reached EOF with unbalanced braces.  It might even be "simple" to support it the same way.

**Comment by Smjert** [comment]
Thanks for the attention on the issue. On my side i'll fix the braces, i saw that it works... unfortunately is a very old code so it's riddled by things like that ^^'.

**Comment by pragmaware** [comment]
Forcing a second pass in this case is pretty easy: see https://github.com/pragmaware/ctags/commit/2972067276dfd6eefb3eea02784ad6e27a88deae .

However it seems that the preprocessor code chooses the same branches also in the second pass and I can't figure out why. Or better, I can't exactly figure out how the "BraceFormat" thingie should work inside the preprocessor code...

**Comment by masatake** [comment]
I'm afraid I injected a bug in upper layer(cprepro and parse.c).

**Comment by masatake** [comment]
I will inspect when I find tme.

**Comment by pragmaware** [comment]
Fixed via #1226
